lib/src/http.dart

```diff
@@ -343,58 +231,202 @@ Map parseJsonResponse(http.Response response) {
+      int maxAttempts = 8,
+      FutureOr<bool> Function(Exception, StackTrace)? retryIf,
+      FutureOr<void> Function(Exception e, int retryCount, http.Request request,
+              http.StreamedResponse? response)?
+          onRetry}) async {
+    late http.Request request;
+    http.StreamedResponse? managedResponse;
+
+    return await retry(() async {
+      final resource = await _httpPool.request();
+      request = await composeRequest();
+
+      http.StreamedResponse directResponse;
+      try {
+        directResponse = await send(request);
+        directResponse.throwIfError();
+      } catch (_) {
+        resource.release();
+        rethrow;
+      }
+
+      // [PoolResource] has no knowledge of streams, so we must manually release
+      // the resource once the stream is canceled or done. We pipe the response
+      // stream through a [StreamController], which enables us to set up lifecycle
+      // hooks.
+      var didStreamActivate = false;
+      final responseController = StreamController<List<int>>(
+        sync: true,
+        onListen: () => didStreamActivate = true,
+      );
+
+      // TODO: does this need to be in a try block? this internally calls
+      // [addStream], which theoretically(?) could throw if the stream has an
+      // error.
+      unawaited(directResponse.stream.pipe(responseController));
+      unawaited(responseController.done.then((_) => resource.release()));
+      managedResponse =
+          directResponse.replacingStream(responseController.stream);
+
+      try {
+        return await onResponse(managedResponse!);
+      } catch (_) {
+        rethrow;
+      } finally {
+        if (!didStreamActivate) {
+          // Release resource if the stream was never subscribed to.
+          unawaited(responseController.stream.listen(null).cancel());
+        }
+      }
+    }, mapException: (e, stackTrace) {
+      if (e is SocketException) {
+        final osError = e.osError;
+        if (osError == null) return PubConnectException('', couldRetry: true);
+
+        // Handle error codes known to be related to DNS or SSL issues. While it
+        // is tempting to handle these error codes before retrying, saving time
+        // for the end-user, it is known that DNS lookups can fail intermittently
+        // in some cloud environments. Furthermore, since these error codes are
+        // platform-specific (undocumented) and essentially cargo-culted along
+        // skipping retries may lead to intermittent issues that could be fixed
+        // with a retry. Failing to retry intermittent issues is likely to cause
+        // customers to wrap pub in a retry loop which will not improve the
+        // end-user experience.
+
+        // TODO: should these return PubConnectException('msg', couldRetry: false)?
+        if (osError.isDnsError) {
+          fail('Could not resolve URL "${request.url.origin}".');
```

Comment:
Maybe retry on dns errors... In EC2 I found them to be intermittent... Have we done any research on this?

Like how long time will the OS cache a failed DNS result?

Reply:
I can look into this. For this first draft, I've left the behavior as close to the original's as possible.